### PR TITLE
Make reservation ID generation backward-compatible with Marathon versions < 1.6

### PR DESCRIPTION
**Pull request by ichernetsky**
JIRA issues: https://jira.mesosphere.com/browse/MARATHON-8032

### CODE REVIEW

File: src/main/scala/mesosphere/marathon/core/instance/Reservation.scala [hunk 1]
```diff
@@ -12,6 +12,34 @@ import play.api.libs.json._
...
+    * Reservation.Id represents a reservation ID which is used in a task reservation labels.
+    */
+  case class Id(runSpecId: String, separator: String, prefix: Option[String], uuid: String) {
+    override def toString: String = prefix match {
+      case Some(prefixStr) => runSpecId + separator + prefixStr + uuid
+      case None => runSpecId + separator + uuid
+    }
+
+    def instanceId: Instance.Id = prefix match {
+      case Some(prefixStr) => Instance.Id(runSpecId + separator + prefixStr + uuid)
```
Comment (alenkacz) [inline src/main/scala/mesosphere/marathon/core/instance/Reservation.scala]: I don't think that reservation id class should have such a deep knowledge of how instance id is concatenated from various string. I think this should definitely be somewhere near Instance.Id implementation

Comment (ichernetsky) [reply src/main/scala/mesosphere/marathon/core/instance/Reservation.scala]: It is implied by the definition of `Instance.Id`.
```
case class Id(idString: String) extends Ordered[Id]
```
It is scattered all over the code. Though I agree, it probably warrants a refactoring.

File: src/main/scala/mesosphere/marathon/core/instance/Reservation.scala [hunk 2]
```diff
@@ -12,6 +12,34 @@ import play.api.libs.json._
...
+      case None => Instance.Id(runSpecId + separator + "marathon-" + uuid)
+    }
+  }
+
+  object Id {
+    private val ReservationIdWithInstanceIdRegex = """^(.+)\.(instance-|marathon-)([^_\.]+)$""".r
```
Comment (jeschkies) [inline src/main/scala/mesosphere/marathon/core/instance/Reservation.scala]: Why not reuse the pattern from `Task.Id`?

**Comment by meln1k** [comment]
I wasn't able to test it locally due to my misbehaving mesos. If someone could try this test case, it would be awesome.

- Start the app from C.1 using Marathon 1.5
- Upgrade Marathon.
- Verify that the app is up and running, and Marathon has not taken any action towards it, and that the file contains only one line.

**Comment by ichernetsky** [comment]
I verified all the C.* test cases yesterday.

File: src/main/scala/mesosphere/marathon/core/instance/Reservation.scala (hunk 2 above)
Comment (ichernetsky) [reply src/main/scala/mesosphere/marathon/core/instance/Reservation.scala]: It is different here.

**Pull request merged**